### PR TITLE
Fix chunked publish with non-block

### LINKED ISSUE #192: Publish using callback does not work with non-blocking mode.
The publish callback is used with `MqttClient_Publish_ex` and it currently uses `MQTT_CODE_CONTINUE` to indicate that the callback should process another block. The problem is that `MQTT_CODE_CONTINUE` is the return value from the non-block read / write to indicate the I/O should be called again. A new return code is needed to specify the condition that the callback should be executed again.

Reference: https://www.wolfssl.com/forums/topic1634-difficulty-using-mqttclientpublishex-with-pic32-harmony-300.html

### PR DESCRIPTION
**Pull request by embhorn**
Created `MQTT_CODE_PUB_CONTINUE` to be used when publishing a payload larger than the MQTT buffer size (publish in chunks). 

```
do {
    rc = MqttClient_Publish(client, publish);
}while ((rc != MQTT_CODE_PUB_CONTINUE) && (rc != MQTT_CODE_CONTINUE));
```

This fixes an issue from ZD12608

This fixes #192

### CODE REVIEW

**Review by dgarske — APPROVED** [review]

File: examples/mqttexample.c [hunk 1]
```diff
@@ -622,3 +622,67 @@ int mqtt_tls_cb(MqttClient* client)
...
+    file = fopen(filePath, "rb");
+    if (file == NULL) {
+        PRINTF("File %s does not exist!", filePath);
+        rc = EXIT_FAILURE;
+        goto exit;
+    }
+
+    /* Determine length of file */
+    fseek(file, 0, SEEK_END);
+    *fileLen = (int) ftell(file);
```
Comment (dgarske) [inline examples/mqttexample.c:648]: Please check this. Could return negative?

File: examples/mqttexample.c [hunk 2]
```diff
@@ -622,3 +622,67 @@ int mqtt_tls_cb(MqttClient* client)
...
+    fseek(file, 0, SEEK_SET);
+    //PRINTF("File %s is %d bytes", filePath, *fileLen);
```
Comment (dgarske) [inline examples/mqttexample.c:650]: Feel free to put this back with `DEBUG_WOLFMQTT`.

File: examples/mqttexample.c [hunk 3]
```diff
@@ -622,3 +622,67 @@ int mqtt_tls_cb(MqttClient* client)
...
+
+    /* Allocate buffer for image */
+    *fileBuf = (byte*)WOLFMQTT_MALLOC(*fileLen);
+    if (*fileBuf == NULL) {
+        PRINTF("File buffer malloc failed!");
+        rc = EXIT_FAILURE;
```
Comment (dgarske) [inline examples/mqttexample.c:656]: Can we use a better return code?

**Review by dgarske — CHANGES_REQUESTED** [review]
This is a great test addition. Just a few minor feedback items.

File: examples/mqttexample.h [hunk 4]
```diff
@@ -74,8 +74,8 @@
 #define DEFAULT_MESSAGE         "test"
 
 #ifdef WOLFMQTT_V5
-#define DEFAULT_MAX_PKT_SZ      768 /* The max MQTT control packet size the
-                                       client is willing to accept. */
+#define DEFAULT_MAX_PKT_SZ      1024*1024 /* The max MQTT control packet size
```
Comment (dgarske) [inline examples/mqttexample.h:77]: Can we also allow a user to override this?

File: examples/mqttexample.c (hunk 3 above)
Comment (dgarske) [reply examples/mqttexample.c:672]: Can we improve the error codes returned in this function?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: examples/mqttexample.h (hunk 4 above)
Comment (embhorn) [reply examples/mqttexample.h:77]: From the example CLI:
`-P <num>    Max packet size the client will accept, default: 1048576`

**Comment by embhorn** [comment]
Weird... I must have undone that return code fix when I changed something else in there.

File: examples/mqttexample.c [hunk 5]
```diff
@@ -622,3 +619,86 @@ int mqtt_tls_cb(MqttClient* client)
...
+            *fileBuf = NULL;
+        }
+    }
+    return rc;
+
+#else
+    (void)filePath;
+    (void)fileBuf;
+    (void)fileLen;
+    #warning No filesystem, so need way to load example firmware file to publish
```
Comment (dgarske) [inline examples/mqttexample.c:701]: The `#warning` isn't valid for Windows VS needs `#pragma message("")`. Note I would prefer not to throw a build warning here. Consider returning an error code instead and let it be a run-time failure.

```
./configure --enable-mqtt5 --enable-nonblock CFLAGS="-DNO_FILESYSTEM" && make

examples/mqttexample.c:701:6: error: No filesystem, so need way to load example firmware file to publish [-Werror,-W#warnings]
    #warning No filesystem, so need way to load example firmware file to publish
     ^
examples/mqttexample.c:701:6: error: No filesystem, so need way to load example firmware file to publish [-Werror,-W#warnings]
    #warning No filesystem, so need way to load example firmware file to publish
     ^
1 error generated.
examples/firmware/fwpush.c:81:41: error: unused parameter 'publish' [-Werror,-Wunused-parameter]
static int mqtt_publish_cb(MqttPublish *publish) {
                                        ^
1 error generated.
examples/mqttexample.c:701:6: error: No filesystem, so need way to load example firmware file to publish [-Werror,-W#warnings]
    #warning No filesystem, so need way to load example firmware file to publish
     ^
examples/mqttexample.c:701:6: error: No filesystem, so need way to load example firmware file to publish [-Werror,-W#warnings]
    #warning No filesystem, so need way to load example firmware file to publish
     ^
```

File: examples/firmware/fwpush.c [hunk 6]
```diff
@@ -134,70 +134,6 @@ static int mqtt_publish_cb(MqttPublish *publish) {
     return ret;
 }
 
-static int fwfile_load(const char* filePath, byte** fileBuf, int *fileLen)
```
Comment (dgarske) [inline examples/firmware/fwpush.c:137]: Above build error with NO_FILESYSTEM.

```
examples/firmware/fwpush.c:81:41: error: unused parameter 'publish' [-Werror,-Wunused-parameter]
static int mqtt_publish_cb(MqttPublish *publish) {
                                        ^
```

File: examples/mqttclient/mqttclient.c [hunk 7]
```diff
@@ -434,51 +434,71 @@ int mqttclient_test(MQTTCtx *mqttCtx)
...
-        prop->type = MQTT_PROP_TOPIC_ALIAS;
-        prop->data_short = mqttCtx->topic_alias;
+    else {
+        mqttCtx->publish.buffer = (byte*)mqttCtx->message;
+        mqttCtx->publish.total_len = (word16)XSTRLEN(mqttCtx->message);
     }
-#endif
 
-    rc = MqttClient_Publish(&mqttCtx->client, &mqttCtx->publish);
+    if (rc == MQTT_CODE_SUCCESS) {
```
Comment (dgarske) [inline examples/mqttclient/mqttclient.c:448]: Else case is not handled here... no error reported for failure like file doesn't exist or feature not compiled in...

**Review by dgarske — CHANGES_REQUESTED** [review]

File: examples/mqttclient/mqttclient.c (hunk 7 above)
Comment (embhorn) [reply examples/mqttclient/mqttclient.c:448]: That was by design, but I can change it. The failure to open is displayed, but the example continues:

```
./examples/mqttclient/mqttclient -f test
MQTT Client: QoS 0, Use TLS 0
MQTT Net Init: Success (0)
MQTT Init: Success (0)
NetConnect: Host test.mosquitto.org, Port 1883, Timeout 5000 ms, Use TLS 0
MQTT Socket Connect: Success (0)
MQTT Connect: Proto (v5), Success (0)
MQTT Connect Ack: Return Code 0, Session Present 0
MQTT Subscribe: Success (0)
  Topic wolfMQTT/example/testTopic, Qos 0, Return Code 0
File test does not exist!
MQTT Waiting for message...
```

Comment (dgarske) [reply examples/mqttclient/mqttclient.c:448]: If NO_FILESYSTEM is defined then nothing is printed. Maybe replace the `#warning` with a PRINTF? I still think the return code should be printed on error.

**Review by dgarske — APPROVED** [review]
Excellent work. Thanks!

**Pull request merged**